server/client.go

```diff
@@ -896,20 +927,22 @@ func (c *client) readLoop() {
 
 		// Main call into parser for inbound data. This will generate callouts
 		// to process messages, etc.
-		if err := c.parse(b[:n]); err != nil {
-			if dur := time.Since(start); dur >= readLoopReportThreshold {
-				c.Warnf("Readloop processing time: %v", dur)
-			}
-			// Need to call flushClients because some of the clients have been
-			// assigned messages and their "fsp" incremented, and need now to be
-			// decremented and their writeLoop signaled.
-			c.flushClients(0)
-			// handled inline
-			if err != ErrMaxPayload && err != ErrAuthentication {
-				c.Error(err)
-				c.closeConnection(ProtocolViolation)
+		for i := 0; i < len(bufs); i++ {
+			if err := c.parse(bufs[i]); err != nil {
+				if dur := time.Since(start); dur >= readLoopReportThreshold {
+					c.Warnf("Readloop processing time: %v", dur)
+				}
+				// Need to call flushClients because some of the clients have been
+				// assigned messages and their "fsp" incremented, and need now to be
+				// decremented and their writeLoop signaled.
+				c.flushClients(0)
+				// handled inline
+				if err != ErrMaxPayload && err != ErrAuthentication {
+					c.Error(err)
+					c.closeConnection(ProtocolViolation)
```

Comment:
I think something else might be required to close the connections here, but not sure, might be good to confirm, for example sending a malformed protocol from the client side seems that the connection is closed in terms of the logs:

```
[47819] 2020/03/12 14:35:24.916249 [ERR] 127.0.0.1:62254 - cid:2 - Client parser ERROR, state=0, i=0: proto='"asdf"...'
[47819] 2020/03/12 14:35:24.916809 [DBG] 127.0.0.1:62254 - cid:2 - Client connection closed
```

but at the system level the client connection remains in CLOSE_WAIT which I think could leak sockets.

```
netstat -an | grep 62254
tcp4       0      0  127.0.0.1.62254        127.0.0.1.4343         CLOSE_WAIT 
```

Reply:
I see what the problem is, will fix.

Reply:
The issue was that I was setting the "closeConnection" flag prior to enqueuing the websocket CloseMessage, which meant that queueOutbound() would just be a no-op. Since the CloseMessage was not sent to your client, I am guessing that it was not closing its end of the socket.

Reply:
Thanks Ivan, though I'm still getting into an indefinite `CLOSE_WAIT` state after getting an unknown protocol error.  Anyway, I tried to do a basic pub/sub using [this cli based client](https://github.com/vi/websocat) and besides the unknown protocol error not resetting the connection I'm also getting an error when sending `PUB` commands (other protocol commands seem to be fine):

```
# client
websocat -k wss://127.0.0.1:4343
INFO {"server_id":"NCCUQTAGJWEP6DZJ4VXCAJX5KX6JNAYXMNULV4ZOSTGJA5IW2AMM6B37","server_name":"NCCUQTAGJWEP6DZJ4VXCAJX5KX6JNAYXMNULV4ZOSTGJA5IW2AMM6B37","version":"2.1.4","proto":1,"go":"go1.14","host":"0.0.0.0","port":4222,"max_payload":1048576,"client_id":1,"client_ip":"127.0.0.1"} 
ping
PONG
pong
CONNECT {}
+OK
PING
sub hello 90
+OK
pub hello 5
world
-ERR 'Unknown Protocol Operation'

# server
[54636] 2020/03/12 16:21:27.311724 [DBG] 127.0.0.1:63175 - wid:1 - Client connection created
[54636] 2020/03/12 16:21:28.948493 [TRC] 127.0.0.1:63175 - wid:1 - <<- [PING]
[54636] 2020/03/12 16:21:28.948515 [TRC] 127.0.0.1:63175 - wid:1 - ->> [PONG]
[54636] 2020/03/12 16:21:30.037835 [TRC] 127.0.0.1:63175 - wid:1 - <<- [PONG]
[54636] 2020/03/12 16:21:32.312171 [TRC] 127.0.0.1:63175 - wid:1 - <<- [CONNECT {}]
[54636] 2020/03/12 16:21:32.312416 [TRC] 127.0.0.1:63175 - wid:1 - ->> [OK]
[54636] 2020/03/12 16:21:34.528213 [DBG] 127.0.0.1:63175 - wid:1 - Client Ping Timer
[54636] 2020/03/12 16:21:34.528303 [TRC] 127.0.0.1:63175 - wid:1 - ->> [PING]
[54636] 2020/03/12 16:21:40.491676 [TRC] 127.0.0.1:63175 - wid:1 - <<- [SUB hello 90]
[54636] 2020/03/12 16:21:40.491745 [TRC] 127.0.0.1:63175 - wid:1 - ->> [OK]
[54636] 2020/03/12 16:21:45.958852 [TRC] 127.0.0.1:63175 - wid:1 - <<- [PUB hello 5]
[54636] 2020/03/12 16:21:45.958880 [TRC] 127.0.0.1:63175 - wid:1 - <<- [PUB hello 5]
[54636] 2020/03/12 16:21:46.966294 [TRC] 127.0.0.1:63175 - wid:1 - ->> [-ERR Unknown Protocol Operation]
[54636] 2020/03/12 16:21:46.966471 [ERR] 127.0.0.1:63175 - wid:1 - Client parser ERROR, state=30, i=5: proto='""...'
[54636] 2020/03/12 16:21:46.966944 [DBG] 127.0.0.1:63175 - wid:1 - Client connection closed
[54636] 2020/03/12 16:21:46.967131 [TRC] 127.0.0.1:63175 - wid:1 - <-> [DELSUB 90]
```

I'm using the TLS certs from the server: 

```hcl
ws {
  port = 4343
  tls {
    cert_file: "test/configs/certs/server-cert.pem"
    key_file: "test/configs/certs/server-key.pem"
  }  
}
```

Reply:
I am going to try with this client, but that may be an issue in that one because works fine with a test that I have using Go gorilla websocket client and Alberto's websocket lib too. Will update after I try.

Reply:
So it looks like this client suppress the `\r`, which our parser fail when it deals with MSG_END_R, it expects the `\r` to be present (unlike other protos apparently that are tolerant of not having it).

Reply:
And I have verified that I am sending the websocket close message, but it seem that this client is not closing the connection in that case?

Reply:
thanks, yep seems like it might be a bad client in this case, ideally clients will follow the close handshake properly though there is no guarantee...  I think in this case what we ought to do besides sending the close message to the client is to set up a close handshake timeout/deadline of a few seconds so that the client has a chance to go away and if it does not gracefully closes on its own after that, then the server resets the connection arbitrarily.  Once LBs get in the middle too I think CLOSE_WAIT issues like this could become common and this could be one way to prevent that from happening.

[1]: https://websockets.readthedocs.io/en/stable/design.html#closing-handshake
[2]: https://github.com/websockets/ws/blob/cb1c893ce94f3d800cd6e4166e6640bda24d74ab/lib/WebSocket.js#L24

Reply:
@wallyqs The server closes its end of the connection right away after sending both the -ERR and WS close message protocols. This is the regular closeConnection() call. ANY client, not just ws will behave the same if the client is not its end of the connection. Check for yourself, here is some test that demonstrate that a non WS client (bare tcp) that does not close its connection still shows the connection state in CLOSE_WAIT until it closes, then we just see the TIME_WAIT. All that is expected with TCP.

test:
```
func TestCloseWait(t *testing.T) {
	o := DefaultOptions()
	o.Port = 4222
	s := RunServer(o)
	defer s.Shutdown()
	c, err := net.Dial("tcp", "127.0.0.1:4222")
	if err != nil {
		t.Fatalf("Error dialing: %v", err)
	}
	defer c.Close()

	br := bufio.NewReader(c)
	l, _, err := br.ReadLine()
	if err != nil {
		t.Fatalf("Error reading: %v", err)
	}

	// Send CONNECT + PING + SUB
	c.Write([]byte("CONNECT {\"verbose\": false}\r\nPING\r\nSUB foo 1\r\n"))
	// Now send a wrong PUB (missing \r)
	c.Write([]byte("PUB foo 2\r\nok\n"))

	// Loop until we get the error
	for err == nil {
		l, _, err = br.ReadLine()
		fmt.Printf("@@IK: got l=%s err=%v\n", l, err)
	}

	fmt.Printf("@@IK: Wait a bit to show that still in CLOSE_WAIT\n")
	time.Sleep(time.Second)
	fmt.Printf("@@IK: Check netstat now!\n")
	out, err := exec.Command("netstat", "-an").Output()
	if err != nil {
		t.Fatalf("Unable to run netstat: %v", err)
	}
	sc := bufio.NewScanner(strings.NewReader(string(out)))
	for sc.Scan() {
		scl := sc.Text()
		if strings.Contains(scl, "4222") {
			fmt.Printf("@@IK: %s\n", scl)
		}
	}
	time.Sleep(time.Second)
	fmt.Printf("@@IK: Now close client connection\n")
	c.Close()
	fmt.Printf("@@IK: Check netstat now!\n")
	out, err = exec.Command("netstat", "-an").Output()
	if err != nil {
		t.Fatalf("Unable to run netstat: %v", err)
	}
	sc = bufio.NewScanner(strings.NewReader(string(out)))
	for sc.Scan() {
		scl := sc.Text()
		if strings.Contains(scl, "4222") {
			fmt.Printf("@@IK: %s\n", scl)
		}
	}
}
```
Output:
```
=== RUN   TestCloseWait
@@IK: got l=PONG err=<nil>
@@IK: got l=-ERR 'Unknown Protocol Operation' err=<nil>
@@IK: got l= err=EOF
@@IK: Wait a bit to show that still in CLOSE_WAIT
@@IK: Check netstat now!
@@IK: tcp4       0      0  127.0.0.1.4222         127.0.0.1.49725        FIN_WAIT_2 
@@IK: tcp4       0      0  127.0.0.1.49725        127.0.0.1.4222         CLOSE_WAIT 
@@IK: tcp4       0      0  127.0.0.1.4222         *.*                    LISTEN     
@@IK: Now close client connection
@@IK: Check netstat now!
@@IK: tcp4       0      0  127.0.0.1.4222         *.*                    LISTEN     
@@IK: tcp4       0      0  127.0.0.1.4222         127.0.0.1.49725        TIME_WAIT  
--- PASS: TestCloseWait (2.05s)
PASS
```

Reply:
Thanks Ivan, yes going into `TIME_WAIT` and also `CLOSE_WAIT` for some time is expected (and `TIME_WAIT` would fade after roughly 30 secs I think), but for example with the plain nats client connected to 4222 that does not call tcp close and then lingers I'm seeing the following after it causes a protocol violation error:

```
[19975] 2020/03/13 10:04:35.383377 [TRC] 127.0.0.1:50724 - cid:2 - ->> [-ERR Unknown Protocol Operation]
[19975] 2020/03/13 10:04:35.383614 [DBG] 127.0.0.1:50724 - cid:2 - Client connection closed

Fri Mar 13 10:04:36 PDT 2020
tcp4       0      0  127.0.0.1.4222         127.0.0.1.50724        FIN_WAIT_2 
tcp4       0      0  127.0.0.1.50724        127.0.0.1.4222         CLOSE_WAIT 
tcp46      0      0  *.4222                 *.*                    LISTEN  

Fri Mar 13 10:06:35 PDT 2020
tcp4       0      0  127.0.0.1.4222         127.0.0.1.50724        FIN_WAIT_2 
tcp4       0      0  127.0.0.1.50724        127.0.0.1.4222         CLOSE_WAIT 
tcp46      0      0  *.4222                 *.*                    LISTEN   
```

After the `FIN` is sent, there a number of keepalive probes sent to the client every 15 seconds (Go default), and after a few minutes the server does seem to be resetting the connection somehow (line which says `localhost.4222 > localhost.50724: Flags [R]`), after that happens then `CLOSE_WAIT` goes away, and I'm not seeing that happen with this `websocat` client for some reason (could be that internally the Go client networking stack is doing something here that this websocat client is not doing, but not sure atm).

```
10:04:35.383566 IP localhost.4222 > localhost.50724: Flags [F.], seq 326, ack 60, win 6378, options [nop,nop,TS val 2447946443 ecr 2447946443], length 0
10:04:35.383568 IP localhost.50724 > localhost.4222: Flags [.], ack 326, win 6374, options [nop,nop,TS val 2447946443 ecr 2447946443], length 0
10:04:35.383573 IP localhost.4222 > localhost.50724: Flags [F.], seq 326, ack 60, win 6378, options [nop,nop,TS val 2447946443 ecr 2447946443], length 0
10:04:35.383578 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447946443 ecr 2447946443], length 0
10:04:35.383581 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447946443 ecr 2447946443], length 0
10:04:50.422186 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:04:50.422209 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:04:50.422213 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:04:50.422240 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447961443 ecr 2447946443], length 0
10:04:50.422247 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:04:50.422249 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447961443 ecr 2447946443], length 0
10:04:50.422258 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447961443 ecr 2447946443], length 0
10:04:50.422263 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447961443 ecr 2447946443], length 0
10:05:05.560877 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:05.560900 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:05.560904 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:05.560930 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447976547 ecr 2447961443], length 0
10:05:05.560937 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:05.560939 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447976547 ecr 2447961443], length 0
10:05:05.560948 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447976547 ecr 2447961443], length 0
10:05:05.560953 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447976547 ecr 2447961443], length 0
10:05:20.602888 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:20.602912 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:20.602916 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:20.602945 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447991547 ecr 2447976547], length 0
10:05:20.602952 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:20.602954 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2447991547 ecr 2447976547], length 0
10:05:20.602963 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447991547 ecr 2447976547], length 0
10:05:20.602967 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2447991547 ecr 2447976547], length 0
10:05:36.131677 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:36.131699 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:36.131714 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:36.131720 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:36.131749 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448007045 ecr 2447991547], length 0
10:05:36.131764 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448007045 ecr 2447991547], length 0
10:05:36.131769 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448007045 ecr 2447991547], length 0
10:05:36.131772 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448007045 ecr 2447991547], length 0
10:05:51.209723 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:51.209740 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:51.209744 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:05:51.209766 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448022077 ecr 2448007045], length 0
10:05:51.209770 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:05:51.209772 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448022077 ecr 2448007045], length 0
10:05:51.209779 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448022077 ecr 2448007045], length 0
10:05:51.209783 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448022077 ecr 2448007045], length 0
10:06:06.275275 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:06:06.275299 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:06.275305 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:06:06.275331 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448037098 ecr 2448022077], length 0
10:06:06.275340 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:06.275342 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448037098 ecr 2448022077], length 0
10:06:06.275356 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448037098 ecr 2448022077], length 0
10:06:06.275364 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448037098 ecr 2448022077], length 0
10:06:21.368506 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:06:21.368531 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:21.368535 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, length 0
10:06:21.368563 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448052142 ecr 2448037098], length 0
10:06:21.368571 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:21.368574 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, options [nop,nop,TS val 2448052142 ecr 2448037098], length 0
10:06:21.368586 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448052142 ecr 2448037098], length 0
10:06:21.368594 IP localhost.4222 > localhost.50724: Flags [.], ack 60, win 6378, options [nop,nop,TS val 2448052142 ecr 2448037098], length 0
10:06:36.406338 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:36.406390 IP localhost.50724 > localhost.4222: Flags [.], ack 327, win 6374, length 0
10:06:36.406421 IP localhost.4222 > localhost.50724: Flags [R], seq 2363662028, win 0, length 0
10:06:36.406429 IP localhost.4222 > localhost.50724: Flags [R], seq 2363662028, win 0, length 0
```

Reply:
Let's be clear: the server is closing the connection, not sure what more it can do. The point of the test I posted is that connection will stay in CLOSE_WAIT until the client closes its end of the socket. If it doesn't it will stay there until the system limit will force the close.
If the websocat client is not closing, that's another story.
The point is that I am not sure what you would like the server to do (regular or ws client does not change a thing as I said since the server closes the connection the same way). Is there a way in Go to force the close of the connection as opposed to shutdown?

Reply:
Thanks for the clarification, ok that is probably the most that we can do right now at the application level it seems then for now...